fw/ss_skb.c

```diff
@@ -1008,6 +1008,94 @@ ss_skb_chop_head_tail(struct sk_buff *skb_head, struct sk_buff *skb,
 	return 0;
 }
 
+/**
+ * Chop @head and @tail bytes at head and end of the message (skb circular
+ * list), iterating over the list if nessessary.
+ * @return 0 on success and -(error code) on error.
+ *
+ * Message length should be greater than the sum of @head and @trail,
+ * otherwise an error is retuned.
+ */
+int
+ss_skb_list_chop_head_tail(struct sk_buff **skb_list_head, 
+			   size_t head, size_t trail)
+{
+
+	struct sk_buff *skb, *skb_hd;
+	size_t sum;
+	int R;
+	skb = *skb_list_head;
+	if (likely(skb->next == skb))
+		goto single_buff;
+	if (likely(head))
+		goto head_and_probably_trail;
+	if (likely(trail))
+		goto trail_only;
+	/* nothing to chop */
+	return 0;
+
+single_buff:
+	/* There is the only 1 buffer in the list */
+	sum = head + trail;
+	if (WARN_ON(skb->len <= sum))
+		return -EBADMSG;
+	if (sum == 0)
+		/* Nothing to chop */
+		/* This check is mostly for jumps from branches below */
+		return 0;
+	return ss_skb_chop_head_tail(skb, skb, head, trail);
+
+head_and_probably_trail:
+	do {
+		if (likely(skb->len > head)) {
+			R = ss_skb_chop_head_tail(skb, skb, head, 0);
+			if (unlikely(R < 0))
+				return R;
+			if (unlikely(trail)) {
+				head = 0;
+				goto trail_only;
+			}
+			return 0;
+		}
+		head -= skb->len;
+		/* We do not use ss_skb_unlink() here to prevent it
+	 	* to remove the last skb in the list and to skip
+	 	* some actions using our apriori knowledge
+	 	*/
+		skb->next->prev = skb->prev;
+		skb->prev->next = skb->next;
+		*skb_list_head = skb_hd = skb->next;
+		__kfree_skb(skb);
+		skb = skb_hd;
+		if (unlikely(skb->next == skb))
+			goto single_buff;
+	} while (head);
```

Comment:
Each loop you check two conditions - why not to rewrite the look with precondition? Just like it's done in `tfw_tls_chop_skb_rec()`?

Reply:
@krizhanovsky 
The initail version of the func has exactly the procondition.
I changed it to poscondition for optimization purpose.
We enter the loop with knowledge that `head` is positive and I do not want to check it once more time.

Reply:
In `tfw_tls_chop_skb_rec()` we have only one `off` (`head` in your code) check, in the loop condition - I don't see why can't we construct this function in similar way and just add the optimization with one skb

Reply:
Well, I'll try to make smth similar.

Reply:
@krizhanovsky 
It seems we are to check the second condition `skb->next == skb` anyway, at least if we want to route the run to single skb case when appropriate and to keep the skb containing several skbs.

Reply:
@krizhanovsky Again, looking into the  `tfw_tls_chop_skb_rec() ` I found sone strange thing.
```C
	skb = data->skb->prev;
	while (unlikely(skb->len <= tail)) {
		tail -= skb->len;
		ss_skb_unlink(&data->skb, skb);
		if (WARN_ON_ONCE(!data->skb))
			return -EIO;
		__kfree_skb(skb);
		skb = data->skb->prev;
	}
```
It seems the `__kfree_skb(skb);` call should be 2 lines upper, otherwise here is a memory leak on error condition raise.

Reply:
> It seems we are to check the second condition skb->next == skb anyway

As in `tfw_tls_chop_skb_rec()` we are to check the second condition `WARN_ON_ONCE(!skb)` (or `(WARN_ON_ONCE(!data->skb))` in tail loop).

And it is realy the same condition.
The difference is the only that `ss_skb_list_chop_head_tail()` check condition _before_ "the fault" trying to prevent it whereas `tfw_tls_chop_skb_rec()` check yhe condition _after_ "the fault" to produce correct reaction and indication.

Reply:
> Again, looking into the tfw_tls_chop_skb_rec() 

Fixed

Reply:
> It seems the __kfree_skb(skb); call should be 2 lines upper, otherwise here is a memory leak on error condition raise.

Good catch!